From 5caedddf11b741a20c6a78d70f147e4ce5329996 Mon Sep 17 00:00:00 2001
From: Heinrich Schuchardt <xypron.glpk@gmx.de>
Date: Sat, 17 Feb 2018 22:47:58 +0100
Subject: [PATCH 1/1] efi_loader: check parameters of guidcmp()

Let guidcmp() return non-zero if one of the parameters is NULL.
This avoids duplicate coding.

Signed-off-by: Heinrich Schuchardt <xypron.glpk@gmx.de>
---
 include/efi_loader.h          | 9 +++++++++
 lib/efi_loader/efi_boottime.c | 4 ++--
 2 files changed, 11 insertions(+), 2 deletions(-)

diff --git a/include/efi_loader.h b/include/efi_loader.h
index b1999f08c6..3a93145fa8 100644
--- a/include/efi_loader.h
+++ b/include/efi_loader.h
@@ -350,8 +350,17 @@ static inline void ascii2unicode(u16 *unicode, const char *ascii)
 		*(unicode++) = *(ascii++);
 }
 
+/*
+ * Compare GUIDs.
+ *
+ * @g1:		1st GUID to compare
+ * @g2:		2nd GUID to compare
+ * @result:	0 if the GUIDS are the same and the pointers are not NULL
+ */
 static inline int guidcmp(const efi_guid_t *g1, const efi_guid_t *g2)
 {
+	if (!g1 || !g2)
+		return -1;
 	return memcmp(g1, g2, sizeof(efi_guid_t));
 }
 
diff --git a/lib/efi_loader/efi_boottime.c b/lib/efi_loader/efi_boottime.c
index d8dff2c65d..a6c3afee72 100644
--- a/lib/efi_loader/efi_boottime.c
+++ b/lib/efi_loader/efi_boottime.c
@@ -1060,7 +1060,7 @@ static efi_status_t efi_get_drivers(struct efi_object *efiobj,
 
 	/* Count all driver associations */
 	list_for_each_entry(handler, &efiobj->protocols, link) {
-		if (protocol && guidcmp(handler->guid, protocol))
+		if (guidcmp(handler->guid, protocol))
 			continue;
 		list_for_each_entry(item, &handler->open_infos, link) {
 			if (item->info.attributes &
@@ -1078,7 +1078,7 @@ static efi_status_t efi_get_drivers(struct efi_object *efiobj,
 		return EFI_OUT_OF_RESOURCES;
 	/* Collect unique driver handles */
 	list_for_each_entry(handler, &efiobj->protocols, link) {
-		if (protocol && guidcmp(handler->guid, protocol))
+		if (guidcmp(handler->guid, protocol))
 			continue;
 		list_for_each_entry(item, &handler->open_infos, link) {
 			if (item->info.attributes &
-- 
2.14.2

